types/react/index.d.ts

```diff
@@ -183,86 +210,107 @@ declare namespace React {
-        ...children: ReactNode[]): DOMElement<P, T>;
-
-    // Custom components
-
-    function createElement<P extends {}>(
-        type: FunctionComponent<P>,
-        props?: Attributes & P | null,
-        ...children: ReactNode[]): FunctionComponentElement<P>;
-    function createElement<P extends {}>(
-        type: ClassType<P, ClassicComponent<P, ComponentState>, ClassicComponentClass<P>>,
-        props?: ClassAttributes<ClassicComponent<P, ComponentState>> & P | null,
-        ...children: ReactNode[]): CElement<P, ClassicComponent<P, ComponentState>>;
-    function createElement<P extends {}, T extends Component<P, ComponentState>, C extends ComponentClass<P>>(
-        type: ClassType<P, T, C>,
-        props?: ClassAttributes<T> & P | null,
-        ...children: ReactNode[]): CElement<P, T>;
-    function createElement<P extends {}>(
-        type: FunctionComponent<P> | ComponentClass<P> | string,
-        props?: Attributes & P | null,
-        ...children: ReactNode[]): ReactElement<P>;
+    // createElement without variadic children
+    function createElement<C extends ReactType>(
+        type: C,
+        props?: Attributes & ComponentProps<C> | null,
+    ): TypedReactElement<C>;
+    function createElement<C extends ComponentClass>(
+        type: C,
+        props?: ClassAttributes<InstanceType<C>> & ComponentPropsWithoutRef<C> | null
+    ): TypedReactElement<C>;
+
+    // types used for variadic children
+
+    type SingleChildTupleType<T> = Extract<T, any> extends never
+        ? never
+        : ([T] | (
+            undefined extends T ? [] : never
+        ));
+    // ReadonlyArray cannot directly be tuples, so we have to cast them to arrays
+    type MultiChildrenTupleType<T extends ReadonlyArray<any>> =
+        T extends ReadonlyArray<infer E> ? E[] : [];
+    // The Extract<T, any> is to work around a problem when T is directly compared with never
+    // ChildrenTupleType<never> results in never (???) if Extract is not used
+    type ChildrenTupleType<T> = Extract<T, any> extends never
+      ? []
+      : SingleChildTupleType<Exclude<T, ReadonlyArray<any>>> | MultiChildrenTupleType<Extract<T, ReadonlyArray<any>>>;
+    // Check first if 'children' key exists to avoid inferring {} if P is {}
+    type ChildrenPropTupleType<P> = 'children' extends keyof P
+        // we need to do the extends check twice because 'undefined' is
+        // not part of the inferred T if the children are optional
+        ? P extends { children: infer T }
+            ? ChildrenTupleType<T>
+            : P extends { children?: infer T }
+                ? ChildrenTupleType<T | undefined>
+                : []
+        : [];
+
+    // createElement with variadic children
+    // tslint:disable-next-line:adjacent-overload-signatures
+    function createElement<C extends ReactType>(
+        type: C,
+        props?: Attributes & Pick<ComponentProps<C>, Exclude<keyof ComponentProps<C>, 'children'>> | null,
```

Comment:
There is not, and even if I were to declare one, `K` would have to not `extends keyof T`. This would end up requiring that components have a `children` prop so I could exclude them, instead of just unconditionally excluding `children`.